homeassistant/components/mysensors/gateway.py

```diff
@@ -155,67 +166,79 @@ def internal_callback(msg):
+    hass: HomeAssistantType, hass_config: ConfigEntry, gateway: BaseAsyncGateway
+):
     """Load any persistent devices and platforms and start gateway."""
     discover_tasks = []
     start_tasks = []
-    for gateway in gateways.values():
-        discover_tasks.append(_discover_persistent_devices(hass, hass_config, gateway))
-        start_tasks.append(_gw_start(hass, gateway))
+    discover_tasks.append(_discover_persistent_devices(hass, hass_config, gateway))
+    start_tasks.append(_gw_start(hass, gateway))
     if discover_tasks:
         # Make sure all devices and platforms are loaded before gateway start.
         await asyncio.wait(discover_tasks)
     if start_tasks:
         await asyncio.wait(start_tasks)
 
 
-async def _discover_persistent_devices(hass, hass_config, gateway):
+async def _discover_persistent_devices(
+    hass: HomeAssistantType, hass_config: ConfigEntry, gateway: BaseAsyncGateway
+):
     """Discover platforms for devices loaded via persistence file."""
     tasks = []
     new_devices = defaultdict(list)
     for node_id in gateway.sensors:
         if not validate_node(gateway, node_id):
             continue
-        node = gateway.sensors[node_id]
-        for child in node.children.values():
+        node: Sensor = gateway.sensors[node_id]
+        for child in node.children.values():  # child is of type ChildSensor
             validated = validate_child(gateway, node_id, child)
             for platform, dev_ids in validated.items():
                 new_devices[platform].extend(dev_ids)
+    _LOGGER.debug("discovering persistent devices: %s", new_devices)
     for platform, dev_ids in new_devices.items():
-        tasks.append(discover_mysensors_platform(hass, hass_config, platform, dev_ids))
+        discover_mysensors_platform(hass, hass_config, platform, dev_ids)
     if tasks:
         await asyncio.wait(tasks)
 
 
-async def _gw_start(hass, gateway):
+async def gw_stop(hass, gateway: BaseAsyncGateway):
+    """Stop the gateway."""
+    _LOGGER.info("stopping gateway %s", gateway.unique_id)
+    if (
+        hasattr(gateway, "connect_task")
+        and gateway.connect_task is not None
+        and not gateway.connect_task.done()
+    ):
+        gateway.connect_task.cancel()
+    await gateway.stop()
+
+
+async def _gw_start(hass: HomeAssistantType, gateway: BaseAsyncGateway):
     """Start the gateway."""
     # Don't use hass.async_create_task to avoid holding up setup indefinitely.
-    connect_task = hass.loop.create_task(gateway.start())
+    gateway.connect_task = hass.loop.create_task(gateway.start())
 
-    @callback
-    def gw_stop(event):
-        """Trigger to stop the gateway."""
-        hass.async_create_task(gateway.stop())
-        if not connect_task.done():
-            connect_task.cancel()
-
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, gw_stop)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, lambda event: gw_stop(hass, gateway)
```

Comment:
Lambdas will run in the executor. That doesn't work with functions that should run in the event loop.